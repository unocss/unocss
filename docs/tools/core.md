---
title: Core
description: The core engine of UnoCSS without any presets. It can be used as the engine of your own atomic CSS framework.
---

# Core

The core engine of UnoCSS without any presets: `@unocss/core`. It can be used as the engine of your own atomic CSS framework.

## Installation

::: code-group

```bash [pnpm]
pnpm add -D @unocss/core
```

```bash [yarn]
yarn add -D @unocss/core
```

```bash [npm]
npm install -D @unocss/core
```

```bash [bun]
bun add -D @unocss/core
```

:::

## Usage

```ts
import { createGenerator } from '@unocss/core'

const generator = await createGenerator(
  { /* user options */ },
  { /* default options */ }
)

const { css } = await generator.generate(code)
```

## License

- MIT License &copy; 2021-PRESENT [Anthony Fu](https://github.com/antfu)
