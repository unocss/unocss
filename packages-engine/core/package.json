{
  "name": "@unocss/core",
  "type": "module",
  "version": "66.1.0-beta.3",
  "description": "The instant on-demand Atomic CSS engine.",
  "author": "Anthony Fu <anthonyfu117@hotmail.com>",
  "license": "MIT",
  "funding": "https://github.com/sponsors/antfu",
  "homepage": "https://unocss.dev",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/unocss/unocss",
    "directory": "packages-engine/core"
  },
  "bugs": {
    "url": "https://github.com/unocss/unocss/issues"
  },
  "keywords": [
    "unocss",
    "atomic-css",
    "atomic-css-engine",
    "css",
    "tailwind",
    "windicss"
  ],
  "sideEffects": false,
  "exports": {
    ".": {
      "types": "./dist/index.d.mts",
      "default": "./dist/index.mjs"
    }
  },
  "main": "dist/index.mjs",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "unbuild",
    "stub": "unbuild --stub",
    "test:attw": "attw --pack --config-path ../../.attw-esm-only.json"
  },
  "devDependencies": {
    "magic-string": "catalog:",
    "unconfig": "catalog:"
  }
}
