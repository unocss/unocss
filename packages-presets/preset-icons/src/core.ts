import type { IconifyJSON } from '@iconify/types'
import type { IconifyLoaderOptions, UniversalIconLoader } from '@iconify/utils'
import type { CSSObject } from '@unocss/core'
import type { IconsOptions } from './types'
import { loadIcon } from '@iconify/utils/lib/loader/loader'
import { searchForIcon } from '@iconify/utils/lib/loader/modern'
import { encodeSvgForCss } from '@iconify/utils/lib/svg/encode-svg-for-css'
import { definePreset, warnOnce } from '@unocss/core'
import icons from './collections'

const COLLECTION_NAME_PARTS_MAX = 3

export { IconsOptions }
export { icons }

/**
 * API for preset-icons
 */
export interface IconsAPI {
  encodeSvgForCss: typeof encodeSvgForCss
  parseIconWithLoader: typeof parseIconWithLoader
  /**
   * This API only available for preset-icons created on Node.js environment
   */
  createNodeLoader?: () => Promise<UniversalIconLoader | undefined>
}

export function createPresetIcons(lookupIconLoader: (options: IconsOptions) => Promise<UniversalIconLoader>) {
  return definePreset((options: IconsOptions = {}) => {
    const {
      scale = 1,
      mode = 'auto',
      prefix = 'i-',
      warn = false,
      iconifyCollectionsNames,
      collections: customCollections,
      extraProperties = {},
      customizations = {},
      autoInstall = false,
      collectionsNodeResolvePath,
      layer = 'icons',
      unit,
      processor,
    } = options

    const flags = getEnvFlags()

    const loaderOptions: IconifyLoaderOptions = {
      addXmlNs: true,
      scale,
      customCollections,
      autoInstall,
      cwd: collectionsNodeResolvePath,
      // avoid warn from @iconify/loader: we'll warn below if not found
      warn: undefined,
      customizations: {
        ...customizations,
        additionalProps: { ...extraProperties },
        trimCustomSvg: true,
        async iconCustomizer(collection, icon, props) {
          await customizations.iconCustomizer?.(collection, icon, props)
          if (unit) {
            if (!props.width)
              props.width = `${scale}${unit}`
            if (!props.height)
              props.height = `${scale}${unit}`
          }
        },
      },
    }

    let iconLoader: UniversalIconLoader

    return {
      name: '@unocss/preset-icons',
      enforce: 'pre',
      options,
      layers: { icons: -30 },
      api: <IconsAPI>{
        encodeSvgForCss,
        parseIconWithLoader,
      },
      rules: [[
        /^([\w:-]+)(?:\?(mask|bg|auto))?$/,
        async (matcher) => {
          let [full, body, _mode = mode] = matcher as [string, string, IconsOptions['mode']]

          iconLoader = iconLoader || await lookupIconLoader(options)

          const usedProps = {}
          const parsed = await parseIconWithLoader(
            body,
            iconLoader,
            { ...loaderOptions, usedProps },
            iconifyCollectionsNames,
          )

          if (!parsed) {
            if (warn && !flags.isESLint)
              warnOnce(`failed to load icon "${full}"`)
            return
          }

          let cssObject: CSSObject
          const url = `url("data:image/svg+xml;utf8,${encodeSvgForCss(parsed.svg)}")`

          if (_mode === 'auto')
            _mode = parsed.svg.includes('currentColor') ? 'mask' : 'bg'

          if (_mode === 'mask') {
            // Thanks to https://codepen.io/noahblon/post/coloring-svgs-in-css-background-images
            cssObject = {
              '--un-icon': url,
              '-webkit-mask': 'var(--un-icon) no-repeat',
              'mask': 'var(--un-icon) no-repeat',
              '-webkit-mask-size': '100% 100%',
              'mask-size': '100% 100%',
              'background-color': 'currentColor',
              // for Safari https://github.com/elk-zone/elk/pull/264
              'color': 'inherit',
              ...usedProps,
            }
          }
          else {
            cssObject = {
              'background': `${url} no-repeat`,
              'background-size': '100% 100%',
              'background-color': 'transparent',
              ...usedProps,
            }
          }

          processor?.(cssObject, {
            ...parsed,
            icon: parsed.name,
            mode: _mode,
          })

          return cssObject
        },
        { layer, prefix },
      ]],
    }
  })
}

export function combineLoaders(loaders: UniversalIconLoader[]) {
  return (async (...args) => {
    for (const loader of loaders) {
      if (!loader)
        continue
      const result = await loader(...args)
      if (result)
        return result
    }
  }) as UniversalIconLoader
}

export function createCDNFetchLoader(
  fetcher: (url: string) => Promise<any>,
  cdnBase: string,
  cacheMap: Map<string, IconifyJSON | Promise<IconifyJSON>> = new Map(),
): UniversalIconLoader {
  function fetchCollection(name: string) {
    if (!icons.includes(name))
      return undefined
    if (!cacheMap.has(name))
      cacheMap.set(name, fetcher(`${cdnBase}@iconify-json/${name}/icons.json`))
    return cacheMap.get(name)!
  }

  return async (collection, icon, options) => {
    let result = await loadIcon(collection, icon, options)
    if (result)
      return result

    const iconSet = await fetchCollection(collection)
    if (iconSet) {
      // possible icon names
      const ids = [
        icon,
        icon.replace(/([a-z])([A-Z])/g, '$1-$2').toLowerCase(),
        icon.replace(/([a-z])(\d+)/g, '$1-$2'),
      ]
      result = await searchForIcon(iconSet, collection, ids, options)
    }

    return result
  }
}

export function getEnvFlags() {
  // eslint-disable-next-line node/prefer-global/process
  const isNode = typeof process !== 'undefined' && process.stdout && !process.versions.deno
  // eslint-disable-next-line node/prefer-global/process
  const isVSCode = isNode && !!process.env.VSCODE_CWD
  // eslint-disable-next-line node/prefer-global/process
  const isESLint = isNode && !!process.env.ESLINT

  return {
    isNode,
    isVSCode,
    isESLint,
  }
}

export async function parseIconWithLoader(
  body: string,
  loader: UniversalIconLoader,
  options: IconifyLoaderOptions = {},
  safeCollectionsNames: string[] = [],
) {
  let collection = ''
  let name = ''
  let svg: string | undefined

  const allCollections = new Set<string>([
    ...icons,
    ...safeCollectionsNames,
    ...Object.keys(options.customCollections || {}),
  ])

  if (body.includes(':')) {
    [collection, name] = body.split(':')

    if (!allCollections.has(collection))
      return

    svg = await loader(collection, name, options)
  }
  else {
    const parts = body.split(/-/g)
    for (let i = COLLECTION_NAME_PARTS_MAX; i >= 1; i--) {
      collection = parts.slice(0, i).join('-')

      if (!allCollections.has(collection))
        continue

      name = parts.slice(i).join('-')
      svg = await loader(collection, name, options)

      if (svg)
        break
    }
  }

  if (!svg) {
    return
  }

  return {
    collection,
    name,
    svg,
  }
}
