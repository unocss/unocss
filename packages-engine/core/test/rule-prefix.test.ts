import { createGenerator } from '@unocss/core'
import presetWind3 from '@unocss/preset-wind3'
import { describe, expect, it } from 'vitest'

describe('prefix', () => {
  it('preset prefix', async () => {
    const uno = await createGenerator({
      presets: [
        presetWind3({ prefix: 'h-' }),
      ],
      rules: [
        ['bar', { color: 'bar' }, { prefix: 'bar-' }],
      ],
      shortcuts: [
        ['shortcut', 'bar-bar', { prefix: 'bar-' }],
      ],
    })

    const unexpected = [
      'text-red',
      'hover:p4',
      'bar',
      'shortcut',
    ]

    const expected = [
      'h-text-red',
      'hover:h-p4',
      'bar-bar',
      'bar-shortcut',
      'h-container',
      '2xl:h-container',
      'dark:children:hover:h-space-x-4',
      'dark:hover:children:h-space-x-4',
      'dark:hover:children:h-divide-x',
      'group-hover:h-bg-red',
      'group-data-[enabled]:h-bg-green',
    ]

    const { css, matched } = await uno.generate(new Set([
      ...unexpected,
      ...expected,
    ]), { preflights: false })

    expect([...matched].sort()).toEqual(expected.sort())
    expect(css).toMatchSnapshot()
  })

  it('uses first truthy prefix', async () => {
    const uno = await createGenerator({
      presets: [
        presetWind3({
          prefix: ['', 'h-'],
        }),
      ],
    })

    expect((await uno.generate('group-hover:h-bg-red group-data-[enabled]:h-bg-green', { preflights: false })).css).toMatchSnapshot()
  })

  it('generate tagged attributify', async () => {
    const uno = await createGenerator({
      presets: [
        presetWind3({
          prefix: 'h-',
          attributifyPseudo: true,
        }),
      ],
    })

    expect((await uno.generate('group-hover:h-bg-red group-data-[enabled]:h-bg-green', { preflights: false })).css).toMatchSnapshot()
  })

  it('multiple preset prefix', async () => {
    const uno = await createGenerator({
      presets: [
        presetWind3({ prefix: ['h-', 'x-'] }),
      ],
      rules: [
        ['bar', { color: 'bar' }, { prefix: ['bar-', 'foo-', ''] }],
        [/^regex$/, () => ({ color: 'regex' }), { prefix: ['bar-', 'foo-', ''] }],
      ],
      shortcuts: [
        ['shortcut', 'bar-bar', { prefix: ['bar-', 'x-'] }],
      ],
    })

    const unexpected = [
      'text-red',
      'hover:p4',
      'bar',
      'shortcut',
    ]

    const expected = [
      'h-text-red',
      'hover:h-p4',

      'bar-bar',
      'foo-bar',
      'bar',

      'bar-regex',
      'foo-regex',
      'regex',

      'bar-shortcut',
      'x-shortcut',

      'h-container',
      '2xl:h-container',
    ]

    const { css, matched } = await uno.generate(new Set([
      ...unexpected,
      ...expected,
    ]), { preflights: false })

    expect([...matched].sort()).toEqual(expected.sort())
    expect(css).toMatchSnapshot()
  })
})
