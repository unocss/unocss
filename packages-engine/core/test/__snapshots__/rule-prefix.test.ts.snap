// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`prefix > generate tagged attributify 1`] = `
"/* layer: default */
[h-group=""][data-enabled] .group-data-\\[enabled\\]\\:h-bg-green{--un-bg-opacity:1;background-color:rgb(74 222 128 / var(--un-bg-opacity));}
[h-group=""]:hover .group-hover\\:h-bg-red{--un-bg-opacity:1;background-color:rgb(248 113 113 / var(--un-bg-opacity));}"
`;

exports[`prefix > multiple preset prefix 1`] = `
"/* layer: default */
.h-container{width:100%;}
.hover\\:h-p4:hover{padding:1rem;}
.h-text-red{--un-text-opacity:1;color:rgb(248 113 113 / var(--un-text-opacity));}
.bar,
.bar-bar,
.bar-shortcut,
.foo-bar,
.x-shortcut{color:bar;}
.bar-regex,
.foo-regex,
.regex{color:regex;}
.\\32 xl\\:h-container{@media (min-width: 1536px){@media (min-width: 640px){
max-width:640px;
}}}
.h-container{@media (min-width: 640px){
max-width:640px;
}}
.\\32 xl\\:h-container{@media (min-width: 1536px){@media (min-width: 768px){
max-width:768px;
}}}
.h-container{@media (min-width: 768px){
max-width:768px;
}}
.h-container{@media (min-width: 1024px){
max-width:1024px;
}}
.\\32 xl\\:h-container{@media (min-width: 1536px){@media (min-width: 1024px){
max-width:1024px;
}}}
.h-container{@media (min-width: 1280px){
max-width:1280px;
}}
.\\32 xl\\:h-container{@media (min-width: 1536px){@media (min-width: 1280px){
max-width:1280px;
}}}
.\\32 xl\\:h-container{@media (min-width: 1536px){
width:100%;
}}
.h-container{@media (min-width: 1536px){
max-width:1536px;
}}
.\\32 xl\\:h-container{@media (min-width: 1536px){@media (min-width: 1536px){
max-width:1536px;
}}}"
`;

exports[`prefix > preset prefix 1`] = `
"/* layer: default */
.h-container{width:100%;}
.dark .dark\\:children\\:hover\\:h-space-x-4>:not([hidden])~:not([hidden])>*:hover{--un-space-x-reverse:0;margin-left:calc(1rem * calc(1 - var(--un-space-x-reverse)));margin-right:calc(1rem * var(--un-space-x-reverse));}
.dark .dark\\:hover\\:children\\:h-space-x-4>:not([hidden])~:not([hidden]):hover>*{--un-space-x-reverse:0;margin-left:calc(1rem * calc(1 - var(--un-space-x-reverse)));margin-right:calc(1rem * var(--un-space-x-reverse));}
.dark .dark\\:hover\\:children\\:h-divide-x>:not([hidden])~:not([hidden]):hover>*{--un-divide-x-reverse:0;border-left-width:calc(1px * calc(1 - var(--un-divide-x-reverse)));border-right-width:calc(1px * var(--un-divide-x-reverse));}
.h-group[data-enabled] .group-data-\\[enabled\\]\\:h-bg-green{--un-bg-opacity:1;background-color:rgb(74 222 128 / var(--un-bg-opacity));}
.h-group:hover .group-hover\\:h-bg-red{--un-bg-opacity:1;background-color:rgb(248 113 113 / var(--un-bg-opacity));}
.hover\\:h-p4:hover{padding:1rem;}
.h-text-red{--un-text-opacity:1;color:rgb(248 113 113 / var(--un-text-opacity));}
.bar-bar,
.bar-shortcut{color:bar;}
.\\32 xl\\:h-container{@media (min-width: 1536px){@media (min-width: 640px){
max-width:640px;
}}}
.h-container{@media (min-width: 640px){
max-width:640px;
}}
.\\32 xl\\:h-container{@media (min-width: 1536px){@media (min-width: 768px){
max-width:768px;
}}}
.h-container{@media (min-width: 768px){
max-width:768px;
}}
.h-container{@media (min-width: 1024px){
max-width:1024px;
}}
.\\32 xl\\:h-container{@media (min-width: 1536px){@media (min-width: 1024px){
max-width:1024px;
}}}
.h-container{@media (min-width: 1280px){
max-width:1280px;
}}
.\\32 xl\\:h-container{@media (min-width: 1536px){@media (min-width: 1280px){
max-width:1280px;
}}}
.\\32 xl\\:h-container{@media (min-width: 1536px){
width:100%;
}}
.h-container{@media (min-width: 1536px){
max-width:1536px;
}}
.\\32 xl\\:h-container{@media (min-width: 1536px){@media (min-width: 1536px){
max-width:1536px;
}}}"
`;

exports[`prefix > uses first truthy prefix 1`] = `
"/* layer: default */
.h-group[data-enabled] .group-data-\\[enabled\\]\\:h-bg-green{--un-bg-opacity:1;background-color:rgb(74 222 128 / var(--un-bg-opacity));}
.h-group:hover .group-hover\\:h-bg-red{--un-bg-opacity:1;background-color:rgb(248 113 113 / var(--un-bg-opacity));}"
`;
